src/data/energy.ts

```diff
@@ -791,3 +792,147 @@ export const getEnergyWaterUnit = (hass: HomeAssistant): string =>
+  from_grid?: { [start: number]: number };
+  to_battery?: { [start: number]: number };
+  from_battery?: { [start: number]: number };
+  solar?: { [start: number]: number };
+}
+
+interface EnergyConsumptionData {
+  total: { [start: number]: number };
+}
+
+export const getSummedData = memoizeOne(
+  (
+    data: EnergyData
+  ): { summedData: EnergySumData; compareSummedData?: EnergySumData } => {
+    const summedData = getSummedDataPartial(data);
+    const compareSummedData = data.statsCompare
+      ? getSummedDataPartial(data, true)
+      : undefined;
+    return { summedData, compareSummedData };
+  }
+);
+
+const getSummedDataPartial = (
+  data: EnergyData,
+  compare?: boolean
+): EnergySumData => {
+  const statIds: {
+    to_grid?: string[];
+    from_grid?: string[];
+    solar?: string[];
+    to_battery?: string[];
+    from_battery?: string[];
+  } = {};
+
+  for (const source of data.prefs.energy_sources) {
+    if (source.type === "solar") {
+      if (statIds.solar) {
+        statIds.solar.push(source.stat_energy_from);
+      } else {
+        statIds.solar = [source.stat_energy_from];
+      }
+      continue;
+    }
+
+    if (source.type === "battery") {
+      if (statIds.to_battery) {
+        statIds.to_battery.push(source.stat_energy_to);
+        statIds.from_battery!.push(source.stat_energy_from);
+      } else {
+        statIds.to_battery = [source.stat_energy_to];
+        statIds.from_battery = [source.stat_energy_from];
+      }
+      continue;
+    }
+
+    if (source.type !== "grid") {
+      continue;
+    }
+
+    // grid source
+    for (const flowFrom of source.flow_from) {
+      if (statIds.from_grid) {
+        statIds.from_grid.push(flowFrom.stat_energy_from);
+      } else {
+        statIds.from_grid = [flowFrom.stat_energy_from];
+      }
+    }
+    for (const flowTo of source.flow_to) {
+      if (statIds.to_grid) {
+        statIds.to_grid.push(flowTo.stat_energy_to);
+      } else {
+        statIds.to_grid = [flowTo.stat_energy_to];
+      }
+    }
+  }
+
+  const summedData: EnergySumData = {};
+  Object.entries(statIds).forEach(([key, subStatIds]) => {
+    const totalStats: { [start: number]: number } = {};
+    const sets: { [statId: string]: { [start: number]: number } } = {};
+    subStatIds!.forEach((id) => {
+      const stats = compare ? data.statsCompare[id] : data.stats[id];
+      if (!stats) {
+        return;
+      }
+
+      const set = {};
+      stats.forEach((stat) => {
+        if (stat.change === null || stat.change === undefined) {
+          return;
+        }
+        const val = stat.change;
+        // Get total of solar and to grid to calculate the solar energy used
+        totalStats[stat.start] =
+          stat.start in totalStats ? totalStats[stat.start] + val : val;
+      });
+      sets[id] = set;
+    });
+    summedData[key] = totalStats;
+  });
+
+  return summedData;
+};
+
+export const computeConsumptionData = memoizeOne(
```

Comment:
These functions are somewhat duplicative of code that seems to already exist elsewhere, but I think we should probably work to try to get more energy cards to use a shared/standardized logic for computing things like consumption. 

Right now we're doing it several different ways in several different cards, even though they're really all computing the same thing with the same source data.